### PR TITLE
Pia 2838 help screen updates

### PR DESCRIPTION
**Pull request by chriskryn**
- view controllers updates
- data sources
- file structures
- design

### CODE REVIEW

File: BankSDK/GiniBankSDKExample/GiniBankSDKExample/ComponentAPI/ComponentAPICoordinator.swift [hunk 1]
```diff
@@ -226,10 +226,13 @@ extension ComponentAPICoordinator {
     fileprivate func showNoResultsScreen() {
         let vc: UIViewController
         if pages.type == .image {
-            let imageAnalysisNoResultsViewController = ImageAnalysisNoResultsViewController()
-            imageAnalysisNoResultsViewController.didTapBottomButton = { [unowned self] in
+            let imageAnalysisNoResultsViewController = HelpTipsViewController(giniConfiguration: giniBankConfiguration.captureConfiguration())
```
Comment (zladzeyka) [inline BankSDK/GiniBankSDKExample/GiniBankSDKExample/ComponentAPI/ComponentAPICoordinator.swift:229]: @chriskryn  I'm wondering why we need this change. Maybe it is better to revert it and comment out broken lines until the new NoResultsViewController will be implemented. What do you think?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpBaseDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        fatalError()
+    }
+    
+    public func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
+    }
+    
+    @objc(tableView:willDisplayCell:forRowAtIndexPath:) public func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+        if items.count == 1 {
+          cell.round(corners: [.bottomLeft, .bottomRight, .topLeft, .topRight], withRadius: 8)
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpBaseDataSource.swift:44]: what about adding a radius corner as a constant?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpMenu/HelpMenuDataSource.swift [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    }()
+    
+    weak var delegate: HelpMenuDataSourceDelegate?
+
+    override init(
+        configuration: GiniConfiguration
+    ) {
+        super.init(configuration: configuration)
+        self.items.append(contentsOf: defaultItems)
+        self.items.append(contentsOf: giniConfiguration.customMenuItems)
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpMenu/HelpMenuDataSource.swift:37]: Should I pick custom items from configuration or giniConfiguration?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpImportViewController.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        configureTableView()
+        configureConstraints()
+        edgesForExtendedLayout = []
+    }
+    
+    public func configureMainView() {
+        self.title = "How to import"
+        view.addSubview(tableView)
+        view.backgroundColor = UIColor.from(giniColor: giniConfiguration.helpScreenBackgroundColor)
+        edgesForExtendedLayout = []
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpImportViewController.swift:52]: @chriskryn you can remove line 52 - you're doing it already in 45

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpImportViewController.swift [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        case .importToApp:
+            cell.headerLabel.text = "2. Import to app"
+            cell.descriptionLabel.text = "Please select your Banking App from the list to start the analysis and transfer process."
+            cell.importImageView.image = UIImageNamedPreferred(named: "helpImport2")
+        }
+        cell.backgroundColor = UIColor.clear
+        cell.headerLabel.backgroundColor = UIColor.clear
+        cell.descriptionLabel.backgroundColor = UIColor.clear
+        cell.contentView.backgroundColor = UIColor.clear
+    }
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpImportViewController.swift:106]: The empty line is missing 
image

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpImportViewController.swift [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if let cell = tableView.dequeueReusableCell(withIdentifier: HelpImportCell.reuseIdentifier) as? HelpImportCell {
+            configureCell(cell: cell, indexPath: indexPath)
+            return cell
+        }
+        fatalError()
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpImportViewController.swift:112]: @chriskryn would you mind adding a message for this fatal error 
image

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpMenuViewController.swift [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+//
+
+import UIKit
+
+/**
+ The `HelpMenuViewControllerDelegate` protocol defines methods that allow you to handle table item selection actions.
+ 
+ - note: Component API only.
+ */
+
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpMenuViewController.swift:16]: @chriskryn could you please add TODO - REMOVE Componen API because we will remove all related Component API code in the nearest future?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpMenuViewController.swift [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+final public class HelpMenuViewController: UIViewController {
+    
+    public weak var delegate: HelpMenuViewControllerDelegate?
+
+    private (set) var dataSource: HelpMenuDataSource
+    private let giniConfiguration: GiniConfiguration
+    private let tableRowHeight: CGFloat = 44
+
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpMenuViewController.swift:33]: Extra empty line :)

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Screen API Coordinator/GiniScreenAPICoordinator+Analysis.swift [hunk 9]
```diff
@@ -23,8 +23,8 @@ extension GiniScreenAPICoordinator {
 // MARK: - ImageAnalysisNoResults screen
 
 extension GiniScreenAPICoordinator {
-    func createImageAnalysisNoResultsScreen() -> ImageAnalysisNoResultsViewController {
-        let imageAnalysisNoResultsViewController: ImageAnalysisNoResultsViewController
+    func createImageAnalysisNoResultsScreen() -> HelpTipsViewController {
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Screen API Coordinator/GiniScreenAPICoordinator+Analysis.swift:26]: We need to agree, on whether should we keep the old Help screen for ImageAnalysisNoResultsViewController or we add some TODOs about the NoResults screen.

File: CaptureSDK/GiniCaptureSDK/Tests/GiniCaptureSDKTests/ImageNoResultsViewControllerTests.swift [hunk 10]
```diff
@@ -11,8 +11,8 @@ import XCTest
 final class ImageNoResultsViewControllerTests: XCTestCase {
     
     let viewControllerTitle = "Title"
-    lazy var viewController: ImageAnalysisNoResultsViewController =
-        ImageAnalysisNoResultsViewController(title: self.viewControllerTitle)
+    lazy var viewController: HelpTipsViewController =
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Tests/GiniCaptureSDKTests/ImageNoResultsViewControllerTests.swift:14]: :) TODO and comment it our or use old NoResultsScreen?

File: HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/ComponentAPICoordinator.swift [hunk 11]
```diff
@@ -379,7 +379,7 @@ extension ComponentAPICoordinator: UINavigationControllerDelegate {
         if toVC is CameraViewController &&
             (fromVC is ReviewViewController ||
              fromVC is AnalysisViewController ||
-             fromVC is ImageAnalysisNoResultsViewController) {
+             fromVC is HelpTipsViewController) {
```
Comment (zladzeyka) [inline HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/ComponentAPICoordinator.swift:382]: Please, 🙏 revert this change, we don't touch HealthSDK now ;)

**Review by zladzeyka — CHANGES_REQUESTED** [review]
@chriskryn Great Job 👍 👏 
image

I've found only small things: empty lines, the only 1 line of duplication
and no result screen case - I would recommend keeping the old No Result if it's possible, it will make it much easier to reimplement it ( there are a few comments about it).
An image named XXXmultiple.png should be a pdf.

File: HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/ComponentAPICoordinator.swift (hunk 11 above)
Comment (chriskryn) [reply HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/ComponentAPICoordinator.swift:382]: @zladzeyka ok, I will revert this change and old NoResults screen too

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpBaseDataSource.swift [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            return cell
+        }
+        fatalError()
+    }
+    
+    public func numberOfSections(in tableView: UITableView) -> Int {
+        return 1
+    }
+    
+    @objc(tableView:willDisplayCell:forRowAtIndexPath:) public func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpBaseDataSource.swift:42]: Is it a new style how to add @objc? I've never seen it.

File: HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/ComponentAPICoordinator.swift (hunk 11 above)
Comment (zladzeyka) [reply HealthSDK/GiniHealthSDKExample/GiniHealthSDKExample/ComponentAPICoordinator.swift:382]: @chriskryn thank you 😊 🙏

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpBaseDataSource.swift (hunk 12 above)
Comment (chriskryn) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpBaseDataSource.swift:42]: Good catch! hmm ... It was automatically added by Xcode after some complains. I need to check it maybe it's no longer necessary

Comment (chriskryn) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpBaseDataSource.swift:41]: It seems so, but it's no longer needed it seems redundant. I will remove it

File: BankSDK/GiniBankSDKExample/GiniBankSDKExample/ComponentAPI/ComponentAPICoordinator.swift [hunk 13]
```diff
@@ -472,7 +472,7 @@ extension ComponentAPICoordinator: UINavigationControllerDelegate {
         if toVC is CameraViewController &&
             (fromVC is ReviewViewController ||
                 fromVC is AnalysisViewController ||
-                fromVC is ImageAnalysisNoResultsViewController) {
+                fromVC is HelpTipsViewController) {
             // When going directly from the analysis or from the single page review screen to the camera the pages
```
Comment (zladzeyka) [inline BankSDK/GiniBankSDKExample/GiniBankSDKExample/ComponentAPI/ComponentAPICoordinator.swift:476]: fromVC is ImageAnalysisNoResultsViewController) {

we need to use ImageAnalysisNoResultsViewController as well

**Review by zladzeyka — CHANGES_REQUESTED** [review]
Great, only one tiny thing in https://github.com/gini/gini-mobile-ios/pull/122/files#diff-a17b4714b1a188bc781fb173bd0986867bd92257a9b25eada5febf52b0a1b56a

File: BankSDK/GiniBankSDKExample/GiniBankSDKExample/ComponentAPI/ComponentAPICoordinator.swift (hunk 13 above)
Comment (chriskryn) [reply BankSDK/GiniBankSDKExample/GiniBankSDKExample/ComponentAPI/ComponentAPICoordinator.swift:476]: ok, I missed it.

**Review by zladzeyka — APPROVED** [review]
Looks good!
image

**Comment by chriskryn** [comment]
@zladzeyka  I have applied the colors and texts changes, could you please have a look again at this ? I will merge it if it looks fine

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+typealias HelpFormatsCollectionSection = (title: String,
+    items: [String],
+    itemsImage: UIImage?)
+
+class HelpFormatsDataSource: NSObject  {
+    let giniConfiguration: GiniConfiguration
+    
+    lazy var sections: [HelpFormatsCollectionSection] = {
+        var sections: [HelpFormatsCollectionSection] =  [
+            (NSLocalizedString(
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift:25]: For all localizable resources, we need to use `NSLocalizedStringPreferredFormat` from GiniCaptureUtils. That will give the possibility to override localizable stings from the app bundle :).

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        if let headerView = view as? UITableViewHeaderFooterView {
+            headerView.contentView.backgroundColor = .clear
+            headerView.backgroundView?.backgroundColor = .clear
+            headerView.textLabel?.textColor = UIColorPreferred(named: "subHeadline")
+        }
+    }
+    
+    func tableView(
+        _ tableView: UITableView,
+        cellForRowAt indexPath: IndexPath
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift:135]: Could you please apply the formatting here?

**Review by zladzeyka — CHANGES_REQUESTED** [review]
Very close :)

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+            cell.separatorView.isHidden = false
+        }
+    }
+}
+
+extension HelpFormatsDataSource: UITableViewDataSource {
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return UITableView.automaticDimension
+    }
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift:98]: @chriskryn Could you please add a new line here?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpFormatsDataSource.swift:100]: @chriskryn what do you think should we remove this empty line here?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpMenu/HelpMenuDataSource.swift [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    override init(
+        configuration: GiniConfiguration
+    ) {
+        super.init(configuration: configuration)
+        self.items.append(contentsOf: defaultItems)
+        self.items.append(contentsOf: configuration.customMenuItems)
+    }
+
+    public override func configureCell(cell: HelpMenuCell, indexPath: IndexPath) {
+
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpMenu/HelpMenuDataSource.swift:41]: And remove empty line here  :)

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpTipsDataSource.swift [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+        cell.backgroundColor = UIColorPreferred(named: "systemWhite")
+        cell.descriptionLabel.text = item.details
+        cell.descriptionLabel.textColor = UIColorPreferred(named: "subHeadline")
+        cell.iconImageView.image = UIImageNamedPreferred(named: item.iconName)
+        cell.separatorView.backgroundColor = UIColorPreferred(named: "separator")
+        cell.selectionStyle = .none
+    }
+    
+    public override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpTipsDataSource.swift:62]: I think we can disable selection for UITableView like:
tableView.allowsSelection = false as well. Can we try it?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpTipsViewController.swift [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+        configureMainView()
+        configureTableView()
+        configureConstraints()
+        edgesForExtendedLayout = []
+    }
+
+    public func configureMainView() {
+        view.addSubview(tableView)
+        view.backgroundColor = UIColorPreferred(named: "systemGray06")
+        edgesForExtendedLayout = []
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/ViewControllers/HelpTipsViewController.swift:54]: it's already in line 48

**Review by zladzeyka — CHANGES_REQUESTED** [review]
Thank you @chriskryn 
Almost there, just a few empty lines manipulation and 2 comments left
image

**Comment by zladzeyka** [comment]
I've tested as well:
My findings:
- Icons to be switched
Untitled

- Bouncing should be disabled for `Tips for photos`.

- Section header is too big:
 Simulator Screen Shot - iPhone 13 - 2022-08-10 at 18 39 47
 
- Some color for the cell's background is missing. 
Simulator Screen Shot - iPhone 8 Plus - 2022-08-10 at 18 33 48 Please see light mode below and [design](https://www.figma.com/file/m4OtCvxWxPMeB0Z5k4Aeow/Gini-Photo-Payment-%2F-iOS?node-id=750%3A9151)
Simulator Screen Shot - iPhone 8 Plus - 2022-08-10 at 18 26 50

**Comment by chriskryn** [comment]
Good catch :+1:. Ok, I have fixed it. The color values again didn't correspond to the color templates. I have added new color into assets as helpBackground. I think we probably should think about naming these kind of new colors.

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/Open with tutorial/OpenWithTutorialCollectionCell.swift [hunk 21]
```diff
@@ -9,130 +9,131 @@
...
         return view
     }()
-    
+
     private func updateStepIndicatorCircleColor(stepIndicatorCircle: UIView) {
-        
-        stepIndicatorCircle.layer.borderColor = UIColor.from(giniColor: GiniConfiguration.shared.indicatorCircleColor).cgColor
+
+        stepIndicatorCircle.layer.borderColor = UIColor.from(
+            giniColor: GiniConfiguration.shared.indicatorCircleColor).cgColor
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/Open with tutorial/OpenWithTutorialCollectionCell.swift:46]: Can we use here directly UIColorPreferred(named: "indicatorCircleColor") ?

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/Open with tutorial/OpenWithTutorialCollectionCell.swift [hunk 22]
```diff
@@ -9,130 +9,131 @@
...
         }
     }
-    
+
     override init(frame: CGRect) {
         super.init(frame: frame)
-        
+
         if #available(iOS 13.0, *) {
             backgroundColor = .systemBackground
```
Comment (zladzeyka) [inline CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/Open with tutorial/OpenWithTutorialCollectionCell.swift:111]: @chriskryn  could you please add this to our palette as well
and use UIColorPreferred😉

File: CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpMenu/HelpMenuDataSource.swift (hunk 18 above)
Comment (zladzeyka) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpMenu/HelpMenuDataSource.swift:41]: It is still there 🙈

Comment (zladzeyka) [reply CaptureSDK/GiniCaptureSDK/Sources/GiniCaptureSDK/Core/Screens/Help/DataSources/HelpMenu/HelpMenuDataSource.swift:41]: line 41

**Review by zladzeyka — APPROVED** [review]
image
Great success!
